site/content/entry/citra-progress-report-2017-june.md

```diff
@@ -0,0 +1,34 @@
++++
+date = "2017-06-27T16:49:00-04:00"
+title = "Citra Progress Report - 2017 June"
+tags = [ "progress-report" ]
+author = "anodium"
+forum = 0000
++++
+
+The summer of 2017 has just rolled in, and although we don't have a [summer of code](https://developers.google.com/open-source/gsoc/), the patches continue rolling in regardless. Because the rate of patches being submitted continues to increase, we've decided to switch from a quarterly schedule to a monthly schedule with the progress reports. With that small announcement out of the way, let's dive right into this month's batch.
+<br />
+
+## [OpenGL: Improve accuracy of quaternion interpolation](https://github.com/citra-emu/citra/pull/2729) by [yuriks](https://github.com/yuriks)
+
+To calculates lighting on any given object, the 3DS' GPU interpolates the light quaternion (the quotient of two vectors) with the surface quaternion of that object. There are three main methods to doing so, the **l**inear int**erp**olation (**lerp**), the **q**uadratic **l**inear int**erp**olation (**qlerp**), and the **s**pherical **l**inear int**erp**olation (**slerp**). All this time Citra used a lerp, which, although the fastest, can lead to a lot of distortion at certain rotation angles.
+
+<p style="text-align: center; font-size: small; padding: 1%">
+<img style="padding: 0% 0% 1% 0%" alt="lerp, qlerp, and slerp being compared with a 145 degree rotation" src="/images/entry/citra-progress-report-2017-june/lerp-qlerp-slerp.gif" />
+<br />
+Notice how the plain lerp (magenta line) lags behind the qlerp (red line) and slerp (blue line), and then speeds up to the other side. Whereas the slerp remains at a constant speed through the entire rotation.
+</p>
+
+[yuriks](https://github.com/yuriks) researched and implemented slerp on Citra, and after a long while of work, it turns out that the 3DS uses lerp as well! The bug in Citra was caused by Citra normalizing the quaternions after interpolating them, when the 3DS normalizes them before, which greatly affected the results since interpolation is not commutative. This particular issue sent them down a very deep rabbit hole, only to lead to a red herring. But, at least it was (eventually) fixed!
```

Comment:
Quick background for @anodium:

Quaternion rotation is done by this formula: p = q*vq 

where:

* q is the quaternion representing your rotation and q* is it's conjugate
* v is a 3-vector (as a pure quaternion) representing the original vector
* p is a 3-vector (as a pure quaternion) representing the result of of the rotation

The norm of the above expression is |p| = |q|^2 |v|. You would expect the norm of the result of a rotation to be equal to the norm of the original vector since you're changing the direction of the vector and not its length (i.e.: |p| = |v|). For this to be true you require |q| = 1. Thus you have to normalize your rotation quaternion before applying the rotation operation; this falls out of the mathematics.

---

I would just state that Citra incorrectly didn't normalize the quaternion before using it in a rotation.